### ISSUE TITLE
Enhance the Stack UX

### ISSUE THREAD

**Issue by svivian** [bolt/bolt#3385]
I recently [asked this question on Stack Overflow](http://stackoverflow.com/questions/29239651/what-does-the-stack-do-in-bolt-cms) but didn't get a good answer. What exactly is the stack supposed to do?

When I'm editing some content I can put files and images on the stack, but I can't insert anything from the stack into my content directly. So I need to click the image or link button on the editor to actually insert stuff.

Doesn't that defeat the purpose? Why would I use the stack instead of just using the editor buttons in the first place?

The SO answer did point out that you need the stack to link to files, because the link button only shows the stack and not the file manager. But that could be easily resolved by adding the file browser in there.

#### Edit by @GawainLynch:
Changing this into an enhancement request for 2.3. 

Summary of 'actionable' points by @svivian:
- The Link dialog in CKEditor should have a way to browse the uploaded files not on the stack.
- The stack should allow selecting of a folder to upload files to.
- Add all recent uploaded files to the stack.
- Some way to drag & drop files from the stack straight into the CKEditor.
- Put access to the stack direct in the Image and Link dialogs, without needing to go into a further dialog (e.g. a "select from stack" button next to the "Browse server" button). Not sure how easy this is but I'm pretty sure you can add custom tabs to the dialogs.

**Comment by bobdenotter** [comment]
My precious stack. :'-(

**Comment by svivian** [comment]
Is there a way to make it useful though? i.e. clicking on files in the stack to insert images/links?

**Comment by xiaohutai** [comment]
As far as I am aware: if you have items on your stack, then you'll get an additional button to choose an item from the stack (you might need to refresh in order to update the contents of the stack for the field itself).

![stack](https://cloud.githubusercontent.com/assets/4630335/7201863/9dc84e96-e50c-11e4-940c-f36012938af8.png)

**Comment by svivian** [comment]
OK... but you still have to browse the server from a different place on the page (or a different page) to add the image to the stack, then "select from stack". You may as well just "select from server" in the first place.

And that only applies specifically to image fields, not general content fields where most people would be adding images and link.

Why not simply have a "recent files" section on the "select from server" window?

**Comment by xiaohutai** [comment]
It's currently a simple list with:
- recently uploaded files (automatically via an image field)
- files directly selected from the server (via the stack, in the sidebar)
- files directly uploaded to the stack (via the stack, in the sidebar)

> Why not simply have a "recent files" section on the "select from server" window?
Might be a good idea for non-image fields where there's no dedicated stack button.

**Comment by bobdenotter** [comment]
We'll work on this, but this needs to be made more "actionable".. Simply stating "it has no use" is not very constructive.

**Comment by svivian** [comment]
@bobdenotter it would be helpful if you could let me know **how you intend the stack to actually be used**. If I don't know what it's for, it's hard to say what should be done. Can you give me your workflow?

I may be missing something obvious, in which case the answer would be to improve the documentation. The only information I can find on the site is the slideshow on the home page, and one paragraph on the features page that more-or-less repeats it. Here's what doesn't make sense to me:

> Easy acces to your latest images

It's not all "latest images", only those uploaded via the stack or manually added to the stack, not those uploaded to different folders in "file management". And IIRC images you select from the server are not automatically added to the stack.

And as I explained above it's not exactly "easy access". Adding links and images in content fields requires going through several windows/tabs.

For links it's especially annoying because there is no access to files besides the stack. I open the link dialog, then browse, realise the file I want to link is not on the stack, close all the windows, add it to the stack on the right (browsing the server), then go back into the link dialogs.

> Quickly move files around between contenttypes

Not really sure what this means. What is being moved, and where? I can access files via the "browse server" windows on any contenttype, so I don't see what the stack is doing in regards to that bullet point.

> Upload everywhere

Yes, sort of. You can upload files using the box on the right, but they all go into the same folder (based on the date like `2015-05`). There is no way to organise them into

> Thumbnail view for quick selecting

Which would be fine, but the thumbnails do nothing. They don't select anything. And on the "select from stack" functionality that @xiaohutai posted above, there are no thumbnails.

**Comment by rarila** [comment]
@svivian Do it like me, simply switch it off ;-)

**Comment by bobdenotter** [comment]
@rarila Barbarian! 

@svivian What I always try to do, when developing something for Bolt, is that I make as little assumptions as possible about how people will use something. This means that i'd like to make things in a sort of MVP-like fashion: working, without too much additional functionality. From there, how something evolves depends on how I, other users and clients would like to use it. 
That said, the feature could use some work. And this will be done, when it gets higher priority (because I want it, because of bugs, or because a paying client wants it), or when all issues that are higher on the list are done (which will not be very soon). 

How I use it is really simple: I upload something to the stack, from whatever page I happen to be on, and then I insert it into the page/content/record where I need it. I'm sure this can be improved by allowing drag'n'drop in more places, but in my daily work it suffices. For me, at least. 

Like I said before, I'm not at all against improving this feature, but simply saying "it's useless" is not very constructive.

**Comment by GawainLynch** [comment]
Honestly, it took me a very long time to grok what the stack was for/about… it wasn't until I set my wife loose on a Bolt backend and she started telling me how awesome this "Stack" thing was… I didn't give her a second of Bolt training either (and that wouldn't have been any good at that point anyway).

I think probably the best thing that it points to me is that we value different approaches and types of users, see our [manifesto](https://docs.bolt.cm/manifesto) for the summary on that.

**Comment by svivian** [comment]
@bobdenotter Sorry if it doesn't seem constructive, that's not my intention. The main problem in my view is that uploading via the stack doesn't allow you to organise files, which means I end up with a convoluted workflow of adding files to the stack manually on one side of the page, then inserting them on the other side, which is *worse* than if the stack didn't exist.

Here's an summary of 'actionable' points I've mentioned above.

- The Link dialog in CKEditor should have a way to browse the uploaded files not on the stack.
- The stack should allow selecting of a folder to upload files to.
- Add all recent uploaded files to the stack.
- Some way to drag & drop files from the stack straight into the CKEditor.
- Put access to the stack direct in the Image and Link dialogs, without needing to go into a further dialog (e.g. a "select from stack" button next to the "Browse server" button). Not sure how easy this is but I'm pretty sure you can add custom tabs to the dialogs.

**Comment by GawainLynch** [comment]
The async routing functions for Stack has been broken out into a topic specific class and if someone is up to the task we can enhance in 2.3, and as such moving this to an enhancement request.

**Comment by SahAssar** [comment]
Any of this we still think is achievable in 2.3?

**Comment by SahAssar** [comment]
Closing due to lack of response from stakeholders. If these enhancements are still desirable please reopen (and have a plan for when/how to implement) please reopen :)

**Comment by svivian** [comment]
What kind of response are you expecting? None of the items have been implemented so there's nothing to comment on.

**Comment by GawainLynch** [comment]
@svivian You're not the only stakeholder that @SahAssar is referencing.

Short version is that we have lots of feature demand, but only so many people to actually do the work, code the PRs, write the unit/acceptance tests, and write documentation.

TL;DR: The point was that an enhancement request over a year old with nobody taking ownership doesn't help anyone. So it was closed until someone steps up (I want to work on it, personally, just that I am so very stretched thin).

**Comment by dadaxr** [comment]
Just giving my 2 cents to say I totally agree with @svivian about the 

>  'actionable' points

he mentions.

I have to finish some stuff in the next weeks but i'll try to have a look on that part :)

**Comment by GawainLynch** [comment]
@dadaxr yeah, all parties in this conversation would be happy with some movement on this… I'm not sure if it bugs @svivian or me the most :laughing: 

I am offline next week, but after that feel free to ping me for help/direction!
